File: HW2 - Dynamic Typing and Duck Typing/Task 3&4/task4_python/Task4Soldier.py
from Soldier import Soldier

class Task4Soldier(Soldier):
    def __init__(self):
        Soldier.__init__(self)
        self.coins = 0
        self.defence = 0
    
    def getCoins(self):
        return self.coins

    def addCoins(self):
        self.coins = self.coins + 1

    def useCoins(self, price):
        self.coins = self.coins - price

    def addDefence(self):
        self.defence = self.defence +5

    def displayInformation(self):
        super(Task4Soldier,self).displayInformation()
        print("Defence: {}".format(self.defence))
        print("Coins: {}".format(self.coins))

    def loseHealth(self):
        super(Task4Soldier,self).loseHealth()
        if self.defence>10:
            self.health = self.health+10
            self.defence = self.defence-10
        else:
            self.health = self.health + self.defence
            self.defence = 0
        return True if self.health<=0 else False

File: HW2 - Dynamic Typing and Duck Typing/Task 3&4/Java/Task4Soldier.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */


/**
 *
 * @author Kwan
 */
public class Task4Soldier extends Soldier{
    protected int coins;
    protected int defence;
    
    public Task4Soldier(){
        super();
        this.coins = 0;
        this.defence = 0;
    }
    
    public int getCoins(){
        return this.coins;
    }
    
    public void addCoins(){
        this.coins += 1;
    }
    
    public void useCoins(int price){
        this.coins -= price;
    }
    
    public void addDefence(){
        this.defence += 5;
    }
    
    public void displayInformation(){
        super.displayInformation();
        System.out.printf("Defence: %d.%n", this.defence);
        System.out.printf("Coins: %d.%n", this.coins);
    }
    
    public boolean loseHealth(){
        super.loseHealth();
        if (this.defence>10){
            this.health += 10;
            this.defence -=10;
        }else{
            this.health += this.defence;
            this.defence = 0;
        }
        return this.health<=0;
    }
        
    
}
